File: java/lib/graphlib/src/main/java/graphlib/Node.java
/*
 *  This file is part of OpenStaticAnalyzer.
 *
 *  Copyright (c) 2004-2017 Department of Software Engineering - University of Szeged
 *
 *  Licensed under Version 1.2 of the EUPL (the "Licence");
 *
 *  You may not use this work except in compliance with the Licence.
 *
 *  You may obtain a copy of the Licence in the LICENSE file or at:
 *
 *  https://joinup.ec.europa.eu/software/page/eupl
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the Licence is distributed on an "AS IS" basis,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the Licence for the specific language governing permissions and
 *  limitations under the Licence.
 */

package graphlib;


import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;

public class Node extends AttributeOwner {

	class NodeTypeSet extends HashSet<NodeType> {
		private static final long serialVersionUID = 7379373638545944057L;
	}

	private final String UID;
	private NodeType type;
	private final List<Edge> edges = new LinkedList<Edge>();

	void addEdge(Edge edge) {
		edges.add(edge);
	}

	void deleteEdge(Edge edge) {
		edges.remove(edge);
	}

	public Node(String UID, NodeType nodeType) {
		super();
		this.type = nodeType;
		this.UID = UID;
	}

	public Node() {
		UID = null;
	}

	public NodeType getType() {
		return type;
	}

	void setType(NodeType nodeType) {
		this.type = nodeType;
	}

	public String getUID() {
		return UID;
	}

	public static class NodeType {

		private String type;

		/**
		 * Create empty NodeType
		 */
		protected NodeType() {

		}

		public NodeType(String type) {
			this.type = type;
		}

		public String getType() {
			return type;
		}

		@Override
		public int hashCode() {
			final int prime = 31;
			int result = 1;
			result = prime * result + ((type == null) ? 0 : type.hashCode());
			return result;
		}

		@Override
		public boolean equals(Object obj) {
			if (this == obj)
				return true;
			if (obj == null)
				return false;
			if (getClass() != obj.getClass())
				return false;
			NodeType other = (NodeType) obj;
			if (type == null) {
				if (other.type != null)
					return false;
			} else if (!type.equals(other.type))
				return false;
			return true;
		}

	}


	/**
	 * It returns every out edge of the graph in a List.
	 * 
	 * @return Out edges of the node.
	 */
	public List<Edge> getOutEdges() {
		List<Edge> ret = new ArrayList<Edge>();
		for (Edge edge : edges) {
			ret.add(edge);
		}
		return ret;
	}

	/**
	 * It returns every out edge of the graph in a List with the given EdgeType.
	 * 
	 * @param type Type of the outedges we would like to find.
	 * @return Out edges of the node with the given type.
	 */
	public List<Edge> findOutEdges(Edge.EdgeType type) {
		List<Edge> ret = new ArrayList<Edge>();
		for (Edge edge : edges) {
			if (edge.getType().equals(type)) {
				ret.add(edge);
			}
		}
		return ret;
	}

	/**
	 * It returns every outedge of the graph in a List with the given types.
	 * 
	 * @param typeset Types of the outedges we would like to find.
	 * @return Out edges of the node with the given type.
	 */
	public List<Edge> findOutEdges(Edge.EdgeTypeSet typeset) {
		List<Edge> ret = new ArrayList<Edge>();
		for (Edge edge : edges) {
			if (typeset.contains(edge.getType())) {
				ret.add(edge);
			}
		}
		return ret;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = super.hashCode();
		result = prime * result + ((UID == null) ? 0 : UID.hashCode());
		result = prime * result + ((edges == null) ? 0 : edges.hashCode());
		result = prime * result + ((type == null) ? 0 : type.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (!super.equals(obj))
			return false;
		if (getClass() != obj.getClass())
			return false;
		Node other = (Node) obj;
		if (UID == null) {
			if (other.UID != null)
				return false;
		} else if (!UID.equals(other.UID))
			return false;
		if (edges == null) {
			if (other.edges != null)
				return false;
		} else if (!edges.equals(other.edges))
			return false;
		if (type == null) {
			if (other.type != null)
				return false;
		} else if (!type.equals(other.type))
			return false;
		return true;
	}

}

File: csharp/lib/graph/Graph/Graph/Model/Node.cs
/*
 *  This file is part of OpenStaticAnalyzer.
 *
 *  Copyright (c) 2004-2017 Department of Software Engineering - University of Szeged
 *
 *  Licensed under Version 1.2 of the EUPL (the "Licence");
 *
 *  You may not use this work except in compliance with the Licence.
 *
 *  You may obtain a copy of the Licence in the LICENSE file or at:
 *
 *  https://joinup.ec.europa.eu/software/page/eupl
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the Licence is distributed on an "AS IS" basis,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the Licence for the specific language governing permissions and
 *  limitations under the Licence.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Xml.Linq;
using Graph.Model.Sub;
using Graph.Model.Super;

namespace Graph.Model
{
	public
	class Node
	: AttributeCollection
	, IDumpable
	{
		internal
		DateTime TimeOfCreation { get; set; }

		List< Edge > edges = new List< Edge >( );

		public
		Node( string uniqueId, string type )
		: base( uniqueId, type ) {}

		internal
		DirectedEdge ConnectTo(Node target, string type)
		{
			DirectedEdge _edge = new DirectedEdge( type, this, target );
			edges.Add( _edge );
			target.edges.Add( _edge );
			return _edge;
		}

		internal
		UndirectedEdge Connect(Node other, string type)
		{
			UndirectedEdge _edge = new UndirectedEdge( type, this, other);
			edges.Add( _edge );
			other.edges.Add( _edge );
			return _edge;
		}

		public
		IEnumerable< Edge > Edges
		{
			get
			{
				return edges;
			}
		}

		public
		IEnumerable< Edge > OutEdges
		{
			get
			{
				return
					edges
						.OfType< DirectedEdge >( )
						.Where( _e => _e.Source == this )
						.Union<Edge>( edges.OfType< UndirectedEdge >( ) );
			}
		}

		public
		IEnumerable< Edge > InEdges
		{
			get
			{
				return
					edges
						.OfType< DirectedEdge >( )
						.Where( _e => _e.Target == this )
						.Union<Edge>( edges.OfType< UndirectedEdge >( ) );
			}
		}

		public
		XElement Dump( )
		{
			var _result = new XElement(
				"node",
				new XAttribute( "name", UniqueID ),
				new XAttribute( "type", Type ) );
			foreach ( var _attribute in Attributes )
			{
				_result.Add( _attribute.Dump( ) );
			}
			foreach ( var _edge in OutEdges )
			{
				_result.Add( _edge.Dump( ) );
			}
			return _result;
		}

		public override
		string ToString( )
		{
			if ( Attributes.Any( _a=>_a.Name=="Name" ) )
			{
				return string.Format( "{0} ({2}) of {1}", UniqueID, Type, Attributes.First( _a => _a.Name == "Name" ).Value );
			}
			else
			{
				return string.Format( "{0} of {1}", UniqueID, Type );
			}
		}
	}
}